### PR TITLE
Clarified "order of includes" rules in c++ style guide.

### PR DESCRIPTION
**Pull request by liangfok**
Modified style guide to state that we are going to use whatever `clang-format` enforces.

### CODE REVIEW

**Comment by david-german-tri** [comment]
LGTM  

@liangfok and I talked about this offline.  The distinction between "your library" and "other libraries" is kind of murky anyway, so my feeling is we should just love Big Formatter.  @jwnimmer-tri - care to do a second review, and merge if you agree?

**Comment by jwnimmer-tri** [comment]
I agree in general that fighting clang-format is a bad idea.

I will make time later on to review this and see about the specifics.

**Comment by liangfok** [comment]
@jwnimmer-tri: Your suspicion is correct. The criteria by which `clang-format` decides whether a header is classified as "other libraries' .h" vs. "your project's .h" is whether it is surrounded by quotes or angle brackets. Specifically, header files surrounded by angle brackets are considered to belong to other libraries whereas those in quotes are classified as Drake's.

For example, `clang-format` leave this ordering alone:

```
#include <gtest/gtest.h>
#include "drake/examples/Acrobot/Acrobot.h"
#include "drake/systems/plants/RigidBodySystem.h"
#include "drake/util/eigen_matrix_compare.h"
```
However, it will reorder this:

```
#include "gtest/gtest.h"
#include "drake/examples/Acrobot/Acrobot.h"
#include "drake/systems/plants/RigidBodySystem.h"
#include "drake/util/eigen_matrix_compare.h"
```

To be:

```
#include "drake/examples/Acrobot/Acrobot.h"
#include "drake/systems/plants/RigidBodySystem.h"
#include "drake/util/eigen_matrix_compare.h"
#include "gtest/gtest.h"
```

**Comment by liangfok** [comment]
Related: http://stackoverflow.com/questions/21593/what-is-the-difference-between-include-filename-and-include-filename

**Comment by liangfok** [comment]
Using double quotes around the first include is OK by `clang-format`, which is good:

```
#include "drake/examples/Acrobot/Acrobot.h"

#include <gtest/gtest.h>

#include "drake/systems/plants/RigidBodySystem.h"
#include "drake/util/eigen_matrix_compare.h"
```

Thus, I think we should clarify the style guide to simply state that all non-Drake header files should be surrounded by angle brackets while all Drake header files should be surrounded by double quotes.

**Comment by liangfok** [comment]
@jwnimmer-tri I'm done clarifying and polishing the PR. Please review when time permits.

File: drake/doc/code_style_guide.rst [hunk 1]
```diff
@@ -28,17 +31,15 @@ We now strictly follow the `Google C++ Style Guide <https://google.github.io/sty
 * No dynamic allocation in the inner simulation/control loops.  Code should be still be thread-safe (e.g. be careful with pre-allocations).
 * Classes and methods should be documented using [doxygen](https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html).
 * Embrace templates/C++11 when it makes the code more correct (more clear or more readable also implies more correct).  Minimize template requirements on public interfaces.  Avoid explicit template instantiations in cc files when possible.
+* For the `order of includes <https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes>`_ style rule, surround Drake header files with double quotes and all other header files with angle brackets. Follow the guide as best you can in terms of arranging the ``#include`` statements in the following order: (1) Related header, (2) C library, (3) C++ library, (4) other libraries' .h, (5) your project's .h. Then accept whatever :ref:`clang-format <code-style-tools-clang-format>` enforces.
+
```
Comment (jwnimmer-tri) [inline drake/doc/code_style_guide.rst]: The (1)..(5) list appears to duplicate google-styleguide exactly, and so this list is brittle with respect to changes in that document (it that document gains additional specific advice, or changes its mind, Drake should follow suit).

**Comment by jwnimmer-tri** [comment]
@liangfok Review complete; just the 1 comment.

**Comment by jwnimmer-tri** [comment]
@david-german-tri 
> The distinction between "your library" and "other libraries" is kind of murky anyway...

I don't think this is true in general, nor for Drake, nor for Drake's use of googletest in this instance.  Project use Drake without the superbuild, and so in those cases with Drake by itself, things like googletest are provided extrinsically to Drake, and so should use angle brackets.

**Comment by liangfok** [comment]
@jwnimmer-tri: Thanks for the quick review. I've made the change. I'm going to assign this PR back to you so you can give it a final review and hopefully merge it.

**Comment by jwnimmer-tri** [comment]
LGTM.  Ready to merge once the CI finishes.

**Comment by david-german-tri** [comment]
I'm very skeptical that angle-brackets are the right way to identify external headers.  Neither the bits of C standard quoted in the stackoverflow thread nor the [examples in the style guide itself](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes) seem to support that.

Moreover, I think the example above only looks like good sorting because it contains no system headers.  @liangfok is using clang-format 3.8, and sent me the results of `clang-format -style=google -dump-config` out-of-thread.  It contains:

```
IncludeCategories:
 - Regex:           '^<.*\.h>'
   Priority:        1
 - Regex:           '^<.*'
   Priority:        2
 - Regex:           '.*'
   Priority:        3
```

Under that config, `<gtest/gtest.h>` would be sorted above `<vector>`, no?  That's clearly more wrong than the way `"gtest/gtest.h"` is sorted.

**Comment by david-german-tri** [comment]
As an aside, this also illustrates that we need to recommend a particular version of `clang-format`.  I'll look into how `-style=google` has evolved.  For instance, in version 3.6 (the last available in Ubuntu 14.04 default repositories) there is no `IncludeCategories` rule at all.

**Comment by liangfok** [comment]
Yes, I'm using `clang-format` version 3.8, which is what's installed using `brew` on OS X 10.11.3.

It turns out `clang-format` version 3.8 will only reorder clumps of `#include` statements that are *not* separated by a newline. Thus, we should revert to the original advice of only using angle brackets for system libraries, and double quotes for everything else. However, the `#include` of drake header files should be separated from external library header files by a newline to keep them below the header files from external libraries.

For example, the following is correct:

```
#include "drake/examples/Acrobot/Acrobot.h"

#include <iomanip>
#include <sstream>
#include <vector>
#include "gtest/gtest.h"

#include "drake/systems/plants/RigidBodySystem.h"
#include "drake/util/eigen_matrix_compare.h"
```

The following is incorrect because `#include "test/gtest.h"` will be moved *below* Drake's headers:

```
#include "drake/examples/Acrobot/Acrobot.h"

#include <iomanip>
#include <sstream>
#include <vector>
#include "gtest/gtest.h"
#include "drake/systems/plants/RigidBodySystem.h"
#include "drake/util/eigen_matrix_compare.h"
```
I will now update this PR to indicate the use of double-quotes for external-but-non-system header files and the need for a blank line separating external / system header files from drake header files.

**Comment by david-german-tri** [comment]
Looks great to me!  OK to merge whenever CI will let us merge.

**Comment by patmarion** [comment]
:-1:  I disagree that you should use double quotes to list third party headers.  I recommend using angle brackets for all non-drake headers.  See earlier discussion here:  https://github.com/RobotLocomotion/drake/pull/1762

In practice, I think the important functional distinction between quotes and brackets is that quotes can be used to reference files (headers or source files) using relative paths without adding header search paths with the -I compiler directive.  Another distinction is that quotes and brackets may be used to separate headers into groups for readability and project specific meaning, to be decided by the project.  The standard leaves the actual behavior to be implementation specific, but these are common practices.  At least for gcc, when its documentation refers to system headers, it often means any header installed under /usr/include or /usr/local/include (and other locations), so that means potentially any third party library.

**Comment by liangfok** [comment]
@david-german-tri, I believe @patmarion would prefer to use angle brackets for all non-drake headers like `<gtest/gtest.h>`. Thoughts?

**Comment by liangfok** [comment]
If we use angle brackets for all non-drake headers, we will need to separate system-headers from non-drake headers with a blank space. In other words, the following would be incorrect since `clang-format` will rearrange `#include <gtest/gtest.h>` to be on top of `#include <vector>`:

```
#include "drake/examples/Acrobot/Acrobot.h"

#include <vector>
#include <gtest/gtest.h>

#include "drake/systems/plants/RigidBodySystem.h"
#include "drake/util/eigen_matrix_compare.h"
```

Instead, the correct format will be:

```
#include "drake/examples/Acrobot/Acrobot.h"

#include <vector>

#include <gtest/gtest.h>

#include "drake/systems/plants/RigidBodySystem.h"
#include "drake/util/eigen_matrix_compare.h"
```
The gap between `#include<vector>` and `#include <gtest/gtest.h>` is optional but I think we should include it in the style guide since it will provide greater consistency in terms of separating different classes of includes with a blank space.

**Comment by patmarion** [comment]
The style guide doesn't explain the motivation for include ordering, apart from the first header, which is obviously the right thing to do.  I'm not actually sure what motivates the order suggested in the style guide.  I would have guessed that this is the more common ordering:

1. drake headers
2. third party headers
3. std lib headers

as an example, http://llvm.org/docs/CodingStandards.html#include-style

**Comment by david-german-tri** [comment]
One of the good things about a style guide is that it replaces endless re-litigation of personal preferences, saving everyone time!

In this case, I think the example is pretty clear: the includes from both `base` (i.e. gtest) and `foo` (i.e. drake) are quoted.  It's also compatible with `clang-format` 3.8+ defaults, as Liang's experiments demonstrate.  Earlier versions are agnostic.  

So, I still think this PR is the right thing to do: it clarifies how to make the style guide and standard tools work well together, without creating a superfluous exception.  :+1:

**Comment by patmarion** [comment]
No.  In the google style guide example, "base" and "foo" quite clearly both reside under the example project "google-awesome-project/src" directory hierarchy.  Thus, in our analogy, they are both "drake headers", and gtest is a third party "system" header.

**Comment by david-german-tri** [comment]
Everything in angle brackets in the example is part of standard C/C++ implementations.  `gtest.h` is more analogous to `commandlineflags.h` (C++ packages of broad utility), than to `vector` or `unistd.h` (symbols supplied by the tools themselves).

If you're still not convinced, how about this.  Chromium is a massive open-source project that [follows the Google style guide](https://www.chromium.org/developers/coding-style), with some exceptions not relevant here.  They also have a very nice code search tool.

`gtest.h"`: [7,003 hits](https://code.google.com/p/chromium/codesearch#search/&q=%22gtest.h%5C%22%22&sq=package:chromium&type=cs)

`gtest.h>`: [191 hits, virtually all in src/third_party](https://code.google.com/p/chromium/codesearch#search/&q=%22gtest.h%3E%22&p=2&sq=package:chromium&type=cs)

**Comment by patmarion** [comment]
Chromium follows Google's vendoring approach to third party dependencies.  For example, many of the search results you linked to are for includes that look like `"testing/gtest/include/gtest/gtest.h"`.

**Comment by david-german-tri** [comment]
That is true, but I'm afraid I don't see the relevance.  Perhaps you could point me at a project that follows the Google style guide and routinely uses angle brackets to `#include` something beyond the standard libraries?

**Comment by patmarion** [comment]
Tensorflow follows the google style guide: https://www.tensorflow.org/versions/r0.7/resources/faq.html

It uses `#include <gtest/gtest.h>`: https://gist.github.com/patmarion/d01680b02d7831eabefeb4b605a936b5

I'm not sure that looking to other projects will be a good decider, because we'll find examples on both sides.

**Comment by patmarion** [comment]
> One of the good things about a style guide is that it replaces endless re-litigation of personal preferences, saving everyone time!

I just want to acknowledge that the google style guide webpage describes some guidelines that are beyond the scope of the clang-format tool.  For those topics, I hope we can continue to engage in open discussion to find what's best for our project.  We already have some exceptions clarified in the code_style_guide.rst.

For sure, this project has agreed to adhere to formatting standards enforceable by the clang-format tool, and selected Google as the `BasedOnStyle` for the project's .clang-format file.

As for the quotes vs angle brackets for third party headers question, I think we can chose either option and the .clang-format file does not need to be customized.  The reason why I jumped in to comment is because acceptance of this pull request implies that we reopen https://github.com/RobotLocomotion/drake/pull/1762, and I am opposed to that.  However, it's just my opinion, I recognize it's a matter of style, and I am willing to let it go.

**Comment by david-german-tri** [comment]
The tensorflow example is really useful, thanks!  It convinces me that quotes-versus-brackets-for-third-party-headers is ambiguous under the guide.  IMO, that means we ought to leave it up to the author.  @liangfok, @patmarion, does that sound reasonable?  Most of this PR would still be useful, there would just be a few words to delete.

@patmarion If you're proposing a broader rewrite of `code_style_guide.rst` (which currently says we "strictly follow the Google C++ Style Guide"), maybe a separate PR would be a better place to discuss?

**Comment by liangfok** [comment]
@david-german-tri and @patmarion: PTAL. Since the [order-of-include](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes) rule in the style guide did not specify whether a particular header file should be surrounded by double quotes or angle brackets, I decided to remove mentioning it from the clarification. Instead, I modified it to simply separate each class of `#include` statement by a blank line, which I think will be sufficient for compatibility with `clang-format`.

**Comment by patmarion** [comment]
looks fine to me!

**Pull request merged**